Update logic to sanitize keywords via Fastlane

mokagio opened the pull request:
App Store Connect requires a keyword string that is 100 characters or less with keywords separated by "English comma, Chinese comma, or any combination of the two."

This change makes sure that sanitized keywords use the English comma.

Note that I actually updated the source of truth for the keywords in [GlotPress](https://translate.wordpress.com/projects/simplenote/ios/release-notes/ar/default/), so this script should not be necessary anymore, but I'm pushing this just in case. Also, this is the code I used to generate the new keywords, so it's handy to have it here for reference, at least till we're sure they're good on GlotPress.

### Test

1. Checkout this branch
1. `git revert 222ef611` to bring the keywords back to a non compliant state
2. `bundle exec fastlane sanitize_appstore_keywords`

The result should match this:

```
برنامج simplenote, والملحوظة, والتدوين, والمُدوِّن, والملحوظات, والمزامنة, والمشاركة, وبسيط, والضبط
```

Notice that I didn't update the keywords metadata file. The reason is that, ideally, we'll download up to date keywords from GlotPress when it's time to finalize the release.

You can verify this by via `cd fastlane && ./download_metadata.swift`. The `metadata/ar-SA/keywords.txt` will update because of the changes I made in GlotPress.

### Review
Only one developer required to review these changes, but anyone can perform the review.

### Release
These changes do not require release notes.

pachlava committed:
Run `Trash` UI tests on every commit

Also, moved sensitive information from Fastfile to CircleCI env variables

pachlava committed:
Added missing `)`

charliescheer committed:
Added alert to inform users that biometry requires pin code

pachlava committed:
New scheme with UI tests

And new CircleCI job to run all tests (with approval)

pachlava committed:
Added approval for full UI Tests set

pachlava committed:
Adjusted selected tests in the subset

pachlava committed:
Removed quotes from params passed to UI tests jobs

pachlava committed:
Testing in MD rendering performance is causing flaky fails

pachlava committed:
Removed default values for some parameter in UI tests job

pachlava committed:
Fixed a `sheme` typo causing the issues

pachlava committed:
Dramatically increased some of the sleeps to see if flaky fails are caused by performance

pachlava committed:
Reduced some of the sleeps

charliescheer committed:
Fixed how localized strings were defined for pin lock required alert

charliescheer committed:
Merge pull request #1299 from Automattic/issue/1256-improve-faceid-flow

Alert users that biometric lock requires setting up a pin code

pachlava committed:
Checking even more increased sleep for some of the actions

pachlava committed:
Subtle change in comment to trigger fresh run

pachlava committed:
Another change in comment to interrupt currently running job

pachlava committed:
Checking if `after_each` gets triggered on cancelled builds

pachlava committed:
A change to trigger job cancellation

pachlava committed:
Another change to trigger job cancellation

pachlava committed:
Removed `after_each` that I've added for testing

pachlava committed:
Fake change to trigger build rerun

pachlava committed:
Testing another way to mark taken test accounts

pachlava committed:
Converted hash to string before sending to server

pachlava committed:
Changed data type that stores account in use

pachlava committed:
Reverted back to using integer for account index

pachlava committed:
Switched to keeping account data in JSON

pachlava committed:
Forgot to convert the hash key

pachlava committed:
Using another ENV variable name that better depicts the data type

pachlava committed:
Shortened one method

pachlava committed:
Added a method that cleans up stale test accounts

pachlava committed:
Forgot to use a new method name

pachlava committed:
Commit to interrupt running job

pachlava committed:
A bit improved logging

pachlava committed:
Removed a redundant method

pachlava committed:
More improved logging

pachlava committed:
Change in order to cancel current the jobs

pachlava committed:
Removed extra line breaks

oguzkocer committed:
Merge pull request #1302 from Automattic/release/4.37

Merge 4.37 code freeze to develop

pachlava committed:
Update fastlane/Fastfile

Co-authored-by: Gio Lodi <giovanni.lodi42@gmail.com>

pachlava committed:
Update fastlane/Fastfile

Co-authored-by: Gio Lodi <giovanni.lodi42@gmail.com>

pachlava committed:
Update .circleci/config.yml

Co-authored-by: Gio Lodi <giovanni.lodi42@gmail.com>

pachlava committed:
Update fastlane/Fastfile

Co-authored-by: Olivier Halligon <olivier.halligon@automattic.com>

pachlava committed:
Sanitizing the hash for all test accounts at once VS each in separate

pachlava committed:
Reverted the previous change

pachlava committed:
Update fastlane/Fastfile

Co-authored-by: Olivier Halligon <olivier.halligon@automattic.com>

pachlava committed:
Reverted the reversion and improved it a bit

pachlava committed:
Merge branch 'add-scheme-with-uitests-subset' of https://github.com/Automattic/simplenote-ios into add-scheme-with-uitests-subset

pachlava committed:
Update fastlane/Fastfile

Co-authored-by: Olivier Halligon <olivier.halligon@automattic.com>

pachlava committed:
Indentation correction

pachlava committed:
Playing around with hash submit

pachlava committed:
Moved to the old way of hash submit

mokagio committed:
Replace UI tests helper local framework with remote Swift package

As part of the 2nd HACK week for 2021, I'm working on UI testing shared
infrastructure.

As step towards sharing more code across projects was extracting
general-purpose helpers present in the Simplenote iOS project into a
dedicated framework.

The second was to create a [Swift
package](https://github.com/Automattic/SwiftUITestsHelpers) from that
code.

With this diff, we replace the local framework with its Swift package
version.

pachlava committed:
Update fastlane/Fastfile

Co-authored-by: Gio Lodi <giovanni.lodi42@gmail.com>

pachlava committed:
Made the 'this runs on CI only' message more obvious

pachlava committed:
Improved phrasing of the error message

pachlava committed:
Merge pull request #1292 from Automattic/run-automated-tests-on-every-commit

Run `Trash` UI tests on every commit

pachlava committed:
Update fastlane/Fastfile

Co-authored-by: Gio Lodi <giovanni.lodi42@gmail.com>

pachlava committed:
Merge branch 'develop' into add-scheme-with-uitests-subset

pachlava committed:
Merge pull request #1301 from Automattic/add-scheme-with-uitests-subset

New scheme with a subset of UI tests + test accounts sanitiser

mokagio committed:
Update Fastlane to version 2.185.0 – Routine update

mokagio committed:
Update Fastlane snapshot integration (SnapshotHelper 1.25)

oguzkocer committed:
Merge pull request #1314 from Automattic/release/4.37

Merge new release notes from 4.37 to develop

mokagio committed:
Update UITestsHelpers to latest dev version

mokagio committed:
Update logic to sanitize keywords via Fastlane

mokagio committed:
Update logic to sanitize keywords via Fastlane

mokagio committed:
Remove unnecessary do in what is now a single-line until block

mokagio commented on fastlane/Fastfile:
```diff
@@ -138,16 +138,20 @@ end
-      until keywords.length <= app_store_connect_keywords_length_limit do
-        keywords = keywords.split(separator)[0...-1].join(separator)
-        puts keywords
+      until keywords.length <= app_store_connect_keywords_length_limit
```
The only thing changed in this line is that there isn't the `do` anymore. It's not necessary because the `until` block has only one line. Thanks Rubocop.

The reason it now has only one line is because I removed that debugging `puts` that I didn't notice when submitting the first version of this work.

mokagio commented on fastlane/Fastfile:
```diff
@@ -138,16 +138,20 @@ end
       end
 
-      until keywords.length <= app_store_connect_keywords_length_limit do
-        keywords = keywords.split(separator)[0...-1].join(separator)
```
In the original version, I didn't know about / realize that ASC wants English commas. The code was trimming the keywords length all right, but because it was putting them back together with the Arabic comma, to ASC they would have looked like a single giant keyword 🤦

mokagio replied:
(Why aren't they using an array? _shakes fist at the sky_)

oguzkocer reviewed (approved):
@mokagio I am a little sad to admit that it took me a while to understand what's going on with this code. The line comments, although helpful in providing context, was making it hard to read the changes, but even after turning off the comments, it took me a second to connect the dots.

This ^ might be just my morning slowness, but I wonder while fixing this code, did we maybe stick to the original a bit too much? As in, would it be better to simply replace the illegal separators with the `english_comma`, so there is no more `separator` keyword we are dealing with? In short, the following feels a bit much for just a simple character replacement:

```
when 'ar-SA'
   # Replace Arabic comma with English one as per App Store
   # Connect requirement
   separator = '،'
   keywords = keywords.split(separator).join(english_comma)
else
   # Nothing do do here, but we might have more languages
   # with different commas in the future
end
```

---

As for testing instructions, the 1-2-3 step one went OK. To my eye, the string you pasted and the string generated are the same. `git diff` doesn't agree, but I assume it's an invalid config on my part, so I think it's fine.

> You can verify this by via cd fastlane && ./download_metadata.swift. The metadata/ar-SA/keywords.txt will update because of the changes I made in GlotPress.

I first didn't understand why we did `cd` and not run the command in root, but looks like it just spits out the data to its folder :joy: This did change the `metadata/ar-SA/keywords.txt` and deleted a bunch of unrelated stuff. So, I think this is fine as well.

oguzkocer committed:
Merge pull request #1319 from Automattic/release/4.37

Merge 4.37 to develop

oguzkocer committed:
[Tooling] Fix GitHub releases for intermediate betas

oguzkocer committed:
Check for toolkit updates in Fastlane

oguzkocer committed:
Update wpmreleasetoolkit to 1.2

oguzkocer committed:
Merge pull request #1321 from Automattic/tooling/fix-gh-releases

[Tooling] Fix GitHub releases for intermediate betas

oguzkocer committed:
Apply suggestions from https://github.com/Automattic/simplenote-ios/pull/1321

oguzkocer committed:
Adds release notes for 4.38

oguzkocer committed:
Merge pull request #1322 from Automattic/tooling/fix-gh-releases-nits

Apply suggestions from #1321

oguzkocer committed:
Merge pull request #1323 from Automattic/4.38-release-notes

Adds release notes for 4.38

oguzkocer committed:
Adds missing PROJECT_ROOT_FOLDER to fastlane

oguzkocer committed:
Merge pull request #1324 from Automattic/add-project-root-folder-to-fastfile

Adds missing PROJECT_ROOT_FOLDER to fastlane

oguzkocer committed:
Bump version number

oguzkocer committed:
Release Notes: add new section for next version (4.39)

oguzkocer committed:
Update draft release notes for 4.38.

oguzkocer committed:
Update strings for localization

oguzkocer committed:
Update metadata strings

oguzkocer committed:
Remove PR numbers from 4.38 release notes

oguzkocer committed:
Update metadata strings

mokagio committed:
Update to use new SwiftUITestsHelper framework name: XCUITestHelpers

mokagio committed:
Switch to using SemVer to resolve XCUITestsHelper dependency

mokagio committed:
Replace local UI test helpers with XCUITestHelpers package (#1315)

oguzkocer committed:
Merge pull request #1325 from Automattic/release/4.38

Merge 4.38.0.0 to develop

mokagio committed:
Update Fastlane and its Simulator snapshot helper (#1316)

mokagio committed:
Simplify keywords Arabic comma separator substitution

Hat tip @oguzkocer

See
https://github.com/Automattic/simplenote-ios/pull/1317#pullrequestreview-680651506

mokagio commented:
> This ^ might be just my morning slowness...

Nah, you're right, that code was a bit too convoluted. Using the switch statement would make sense if there were more cases, but with only one it adds cruft.

> In short, the following feels a bit much for just a simple character replacement

Totally. I should have taken a further step backward 😄 This is why it's great to do code reviews, hey!

I'll requested a new review from you to check my simplified code and enabled auto-merge to close the loop on this.

oguzkocer commented:
@mokagio It looks like auto-merge might have merged your changes prematurely. I see that there has been a force push as well, so.. not sure how to proceed with this one.

oguzkocer commented:
@mokagio fwiw, I got a ping in this commit d589edc309cc0976beb6ee5cacb8e43b1daa6fb4 and assuming you've tested it, that commit seems fine to me 🤷

And as I was typing that, I just realized this all got merged to `trunk`, because Github updated it when `release/4.37` was removed 😱 🤦

oguzkocer commented:
Sorry for the 3rd comment in 7 minutes, I just compared trunk and develop and the diff is what this PR is supposed to be. @mokagio I suggest we open a PR from `trunk` to `develop` and merge it in with a PR review. `trunk` is ahead of where it's supposed to be at this time of code freeze, but `trunk` is not a very useful branch, so not a big loss 🙈

mokagio commented:
![](https://camo.githubusercontent.com/d7a49ad2d27ea376ff9ec3855158c3f31d2ac085e3bf16dcaf627b0f63b5e2f9/68747470733a2f2f6d656469612e67697068792e636f6d2f6d656469612f647644304f41455452664358432f67697068792e676966)

Can't believe I opened this against `trunk` and didn't realize it 😢 

I don't want `trunk` to be ahead and I worry that merging the next release into it might be annoying.

What do you think of:

1. Opening a PR to revert this merge into `trunk` (I thought of rewriting the history, should be safe because who's checking out `trunk` anyways, but don't know... a revert seems safer)
2. Open a PR against `develop` (`develop`, Gio! `d e v e l o p`) for this fix?
